src/main/kotlin/io/ionic/libs/ionfilesystemlib/IONFILEController.kt

```diff
@@ -0,0 +1,286 @@
+
+                    else -> contentResolverHelper.readFileInChunks(resolvedUri.uri, options)
+                }
+                emitAll(readByChunksFlow)
+            },
+            onFailure = { throw it }
+        )
+    }
+
+    /**
+     * Get metadata / information on a file or directory
+     *
+     * @param uri a [IONFILEUri] object; will resolve internally to determine the actual file
+     * @return success with a [IONFILEMetadataResult] object containing information on the file, or error otherwise
+     */
+    suspend fun getMetadata(uri: IONFILEUri): Result<IONFILEMetadataResult> =
+        uriHelper.useUriIfResolved(uri) { resolvedUri ->
+            if (resolvedUri is IONFILEUri.Resolved.Local) {
+                localFilesHelper.getFileMetadata(resolvedUri.fullPath)
+            } else {
+                contentResolverHelper.getFileMetadata(resolvedUri.uri)
+            }
+        }
+
+    /**
+     * Save contents to a file
+     *
+     * This method will fail if a "content://" type URI is passed
+     *
+     * @param uri a [IONFILEUri] object; will resolve internally to determine the actual file
+     * @param options data and options for configuring file saving; see [IONFILESaveOptions]
+     * @return success with the android [Uri] that was saved to, or error otherwise
+     */
+    suspend fun saveFile(
+        uri: IONFILEUri,
+        options: IONFILESaveOptions
+    ): Result<Uri> = uriHelper.useUriIfResolvedAsLocalFile(uri) { resolvedLocalFile ->
+        localFilesHelper.saveFile(resolvedLocalFile.fullPath, options)
+            .map { resolvedLocalFile.uri }
+    }
+
+    /**
+     * List the contents of a directory - all children including files and directories
+     *
+     * This method is not recursive, meaning that if the directories has sub-directories, the children of these will not be listed
+     *
+     * This method will fail if a file is passed instead of a directory
+     *
+     * @param uri a [IONFILEUri] object; will resolve internally to determine the actual directory
+     * @return success with list of [IONFILEMetadataResult] objects containing information about directory's children, or error otherwise
+     */
+    suspend fun listDirectory(uri: IONFILEUri): Result<List<IONFILEMetadataResult>> =
+        uriHelper.useUriIfResolvedAsLocalDirectory(uri) { resolvedUri ->
+            directoriesHelper.listDirectory(resolvedUri.fullPath)
+        }
+
+    /**
+     * Delete a file or directory
+     *
+     * @param uri a [IONFILEUri] object; will resolve internally to determine the actual path
+     * @param options the options to configure deletion; only applies for directories; see [IONFILEDeleteOptions]
+     * @return success if file / directory was deleted, error otherwise
+     */
+    suspend fun delete(uri: IONFILEUri, options: IONFILEDeleteOptions): Result<Unit> =
+        uriHelper.useUriIfResolved(uri) { resolvedUri ->
+            when {
```

Comment:
Not super important, but this when could be refactored to:

```
when (resolvedUri) {
    is IONFILEUri.Resolved.Local -> {
        val path = resolvedUri.fullPath
        if (resolvedUri.type == LocalUriType.DIRECTORY) {
            directoriesHelper.deleteDirectory(path, options)
        } else {
            localFilesHelper.deleteFile(path)
        }
    }
    else -> contentResolverHelper.deleteFile(resolvedUri.uri)
}

```

Reply:
Yeah, I guess the only downside with the current approach is that we're duplicating the `resolvedUri is IONFILEUri.Resolved.Local` check, while with the proposed approach we avoid that.

But we can keep it how it is, I don't have a strong preference about this.